### ISSUE TITLE
Bundler 1.16.0 errors on Circle CI with load_path_cache

### ISSUE THREAD

**Issue by ebeigarts** [Shopify/bootsnap#103]
We started receiving errors on CirrleCI, I think they might have updated the docker image `circleci/ruby:2.4-node-browsers` to use newer bundler (1.15 => 1.16) or newer ruby version (2.4.1 => 2.4.2) or both.

This looks like a bug in bootsnap or bundler, because withtout bootsnap everything works ok.

It doesn't work with bootsnap + bundler 1.16 + `load_path_cache: true` on CircleCI.
It works with bootsnap + bundler 1.16 + `load_path_cache: false` on CircleCI.
It works with bootsnap + bundler 1.15 + `load_path_cache: true` on CircleCI.

The problem might be bundler 1.16 + bootsnap, but I can't reproduce this on OSX, only in CircleCI, haven't tried this on other linux machines.

Any ideas how to fix or debug this?

The error when running  `rails c`:

```
/usr/local/lib/ruby/site_ruby/2.4.0/bundler/runtime.rb:84:in `rescue in block (2 levels) in require': There was an error while trying to load the gem 'dotenv-rails'. (Bundler::GemRequireError)
Gem Load Error is: Trying to register Bundler::GemfileError for status code 4 but Bundler::GemfileError is already registered
Backtrace for gem load error is:
/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/errors.rb:9:in `status_code'
/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/errors.rb:20:in `<class:GemfileError>'
/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/errors.rb:20:in `<module:Bundler>'
/usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/errors.rb:3:in `<main>'
/home/circleci/app/vendor/bundle/ruby/2.4.0/gems/bootsnap-1.1.5/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:17:in `require'
/home/circleci/app/vendor/bundle/ruby/2.4.0/gems/bootsnap-1.1.5/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:17:in `require'
```

Full log: https://gist.github.com/ebeigarts/3d4d2c03aec360fecd4132dd062a1589

**Comment by burke** [comment]
Can you confirm that you're loading `bootsnap/setup` after `bundler/setup`? Removing the cache at `tmp/cache/bootsnap-load-path-cache` *may* resolve the issue you're experiencing, though there's clearly a bug here.

**Comment by ebeigarts** [comment]
Yes, `bootsnap/setup` is after `bundler/setup`

config/boot.rb:

```
# frozen_string_literal: true

require 'rubygems'

min_ruby_version = "2.4.1"
if Gem::Version.new(RUBY_VERSION) < Gem::Version.new(min_ruby_version)
  raise "Your Ruby version is #{RUBY_VERSION}, but this application requires #{min_ruby_version}"
end

# Set up gems listed in the Gemfile.
ENV['BUNDLE_GEMFILE'] ||= File.expand_path('../Gemfile', __dir__)

require 'bundler/setup' # Set up gems listed in the Gemfile.
require 'bootsnap/setup' # Speed up boot time by caching expensive operations.
```

**Comment by halo** [comment]
For what it's worth, this happens to me on macOS in Rails. I only upgraded bundler to `1.16.0` and then this occurred.

**Comment by edwardloveall** [comment]
I just worked around this by changing my boot.rb to the following:

```ruby
ENV['BUNDLE_GEMFILE'] ||= File.expand_path('../Gemfile', __dir__)

require 'bundler/setup' # Set up gems listed in the Gemfile.
require 'bootsnap'
env = ENV['RAILS_ENV'] || 'development'
circle_off = ENV['CIRCLECI'].nil?
Bootsnap.setup(
  cache_dir: 'tmp/cache',
  development_mode: env == 'development',
  load_path_cache: circle_off,
  autoload_paths_cache: circle_off,
  disable_trace: true,
  compile_cache_iseq: true,
  compile_cache_yaml: true
)
```

Check checks for the `CIRCLECI` environment variable and turns off `load_path_cache` and `autoload_paths_cache` when using Circle.

**Comment by tommy2118** [comment]
The above boot.rb did not work for me.  I ended up using this and it works.
```ruby
ENV['BUNDLE_GEMFILE'] ||= File.expand_path('../Gemfile', __dir__)

require 'bundler/setup' # Set up gems listed in the Gemfile.
require 'bootsnap/setup' unless ENV['CIRCLECI']
```

**Comment by mattias-lundell** [comment]
Similar experience on travis..

**Comment by bobmaerten** [comment]
Same here with SemaphoreCI :/

**Comment by coolnalu** [comment]
Just happened to us.

**Comment by timothysu** [comment]
This occurs for me locally with Rubygems 2.7.5.  Downgrading to 2.7.4 fixes it:
`gem update --system 2.7.4`
`rvm all do gem update --system 2.7.4`

**Comment by UlyssesInvictus** [comment]
small word to the wise that `disable_trace` breaks debuggers, so caveat emptor if you're just pasting in @edwardloveall 's fix to solve circle issues

**Comment by wellington1993** [comment]
Hi, I removed `require 'bootsnap/setup'` from `boot.rb` and worked for me.

**Comment by jules2689** [comment]
Just an FYI @wellington1993 ... that removes Bootsnap from your application entirely, which is why you saw this work.

We haven't been able to replicate this error at all, so it's been difficult to look into. We just had someone replicate it locally so we should be able to look into what is happening here.

Sorry for the delay.

**Comment by jules2689** [comment]
We've been able to trace this down to an issue caused by full path resolution. Steps to reproduce are:
1) Load bundler
2) Load an alternate bundler explicitly via absolute path

Here is the basic case:
![image](https://user-images.githubusercontent.com/3074765/37052831-86b6f07e-2148-11e8-839a-65509d4159b8.png)

**Comment by burke** [comment]
I have a branch that I believe fixes this error: https://github.com/Shopify/bootsnap/pull/138

Can anyone give Bootsnap version 1.2.0.pre a shot and let me know how this affects things?

**Comment by haberbyte** [comment]
I experienced the same issue in a docker image and with 1.2.0.pre it works fine!! 👍

**Comment by edwardloveall** [comment]
I was having this issue on CircleCI and 1.2.0.pre seems to have fixed it! 👍

**Comment by burke** [comment]
Great! I just released 1.2.0 so I'll close.
